core/src/net/tcp_stream.cpp

```diff
@@ -0,0 +1,165 @@
+{
+    mSock = socket(AF_INET, SOCK_STREAM, 0);
+    if (mSock == InnerInvalidSocket)
+    {
+        CUBOS_ERROR("Failed to allocate TCP socket: {}", logSystemError());
+        return false;
+    }
+
+    sockaddr_in addr;
+    addr.sin_family = AF_INET;
+    addr.sin_addr.s_addr = inet_addr(address.toString().c_str());
+    addr.sin_port = htons(port);
+
+    if (::connect(mSock, (struct sockaddr*)&addr, sizeof(addr)) < 0)
+    {
+        CUBOS_ERROR("Failed to connect TCP stream to {}:{}: {}", address.toString(), port, logSystemError());
+        return false;
+    }
+
+    CUBOS_DEBUG("New TCP stream at address {} and port '{}'", address.toString(), port);
+    return true;
+}
+
+void TcpStream::inner(InnerSocket socket)
+{
+    disconnect();
+    mSock = socket;
+    CUBOS_DEBUG("New TCP stream at socket {}", mSock);
+}
+
+void TcpStream::disconnect()
+{
+    if (mSock != InnerInvalidSocket)
+    {
+        CUBOS_WARN("Closing TCP stream socket: '{}'", mSock);
+#ifdef _WIN32
+        closesocket(mSock);
+        WSACleanup();
+#else
+        close(mSock);
+#endif
+        mSock = InnerInvalidSocket;
+    }
+}
+
+TcpStream::TcpStream()
+{
+#ifdef _WIN32
+    WSADATA wsa;
+    CUBOS_ASSERT(WSAStartup(MAKEWORD(2, 2), &wsa) == 0, " WSAStartup failed: {}", WSAGetLastError());
+#endif
+}
+
+TcpStream::~TcpStream()
+{
+    disconnect();
+}
+
+std::size_t TcpStream::read(void* buffer, std::size_t size)
+{
+    if (mSock == InnerInvalidSocket)
+    {
+        CUBOS_ERROR("Invalid socket");
+        return 0;
+    }
+
+#ifdef _WIN32
+    auto bytesRead = ::recv(mSock, static_cast<char*>(buffer), static_cast<int>(size), 0);
+#else
+    auto bytesRead = ::recv(mSock, buffer, size, 0);
+#endif
+    if (bytesRead < 0)
+    {
+        CUBOS_ERROR("Failed to read from TCP stream: {}", logSystemError());
+        return 0;
+    }
+
+    if (bytesRead == 0)
+    {
+        CUBOS_WARN("TCP stream EOF");
+        mEof = true;
+    }
+
+    CUBOS_INFO("Incoming TCP message");
+    auto bytesReadCasted = static_cast<std::size_t>(bytesRead);
+    mPos += bytesReadCasted;
+    return bytesReadCasted;
+}
+
+std::size_t TcpStream::write(const void* buffer, std::size_t size)
+{
+    if (mSock == InnerInvalidSocket)
+    {
+        CUBOS_ERROR("Invalid socket");
+        return 0;
+    }
+
+#ifdef _WIN32
+    auto bytesWritten = ::send(mSock, static_cast<const char*>(buffer), static_cast<int>(size), 0);
+#else
+    auto bytesWritten = ::send(mSock, buffer, size, 0);
+#endif
+    if (bytesWritten < 0)
+    {
+        CUBOS_ERROR("Failed to write TCP buffer: {}", logSystemError());
+        return 0;
+    }
+
+    CUBOS_INFO("Sent TCP message");
+    auto bytesWrittenCasted = static_cast<std::size_t>(bytesWritten);
+    mPos += bytesWrittenCasted;
+    return bytesWrittenCasted;
+}
+
+char TcpStream::peek() const
+{
+    if (mSock == InnerInvalidSocket)
+    {
+        CUBOS_ERROR("Invalid socket");
+        return ' ';
```

Comment:
```suggestion
        return ' \0';
```
Is probably a more fit 'default' character.